Feature urlpath in ModularServer

Holzhauer opened the pull request:
When deploying several models on a public server behind a proxy, adding a url path segment is required, e.g. http://server/mesa-schelling > localhost:8521/mesa-schelling

rht commented on mesa/visualization/ModularVisualization.py:
```diff
@@ -240,17 +240,22 @@ class ModularServer(tornado.web.Application):
     verbose = True
 
     port = int(os.getenv("PORT", 8521))  # Default port to listen on
+    urlpath = str(os.getenv("URLPATH", ""))
```
Just to be sure, this `urlpath` variable needs to have the trailing `/` in the end to be stripped, otherwise it becomes `/urlpath//local/...` later on.

rht commented:
I'm updating the CONTRIBUTING.rst to include this guide: https://github.com/projectmesa/mesa/pull/1298/commits/86b5095f09ae4dc2a954f97def0d9c4f863b1018. You need to run `git config pull.rebase true` in the Mesa Git repo, and the do a `git pull` from the Mesa upstream repo. With this, you won't see the messy merge commits. The workflow is that you will have to do `git push --force` (or `git push -f`) when pushing to GH remote, because you are rewriting history by rebasing your new commits on top of Mesa main, instead of doing merge.

Holzhauer commented:
Shame on me, indeed passing the `urlpath` param makes no sense after `ModularServer` is instantiated. As it is not required I removed it. Also stripped slashes from `URLPATH` and added some doc.
However, I struggle to run `pre-commit` as I have not python3.6 on my system. Why is `language_version` set to `python3.6` in `.pre-commit-config.yaml` when mesa is built only for python >=3.7? Shouldn't `python3` suffice?
Further, in `CONTRIBUTING.rst`, shouldn't it read `git commit -m "Fix X issue."` instead of `git commit -m "Fixes X issue."` ;-)?

rht commented:
Yeah, the .pre-commit-config.yaml (maybe just remove the `language_version`) and CONTRIBUTING.rst are definitely outdated. A PR to fix them would be appreciated!

rht commented:
> As it is not required I removed it. Also stripped slashes from URLPATH and added some doc.

I think these changes haven't been pushed to GH?

Holzhauer commented:
Is there a reason why handlers are initialised outside __init__() in ModularServer? I wonder because this way it is hard to define a test which requires setting the environment var URLPATH before the initisalisation of handlers.

rht commented:
Yeah, I think they are fine to be moved to `__init__`. But this refactor should be done in a separate PR, so that the PRs remain small and faster to merge.

Holzhauer commented:
OK, could you then merge this PR before I prepare the refactor PR (otherwise there will be conflicts in ModularVisualization.py)?

rht commented on mesa/visualization/templates/modular_template.html:
```diff
@@ -80,22 +81,24 @@ <h4 class="modal-title">About {{ model_name }}</h4>
     <!-- Bottom-load all JavaScript dependencies -->
-    <script src="/static/external/bootstrap-5.1.3-dist/js/bootstrap.bundle.min.js"></script>
-    <script src="/static/external/bootstrap-slider-11.0.2/dist/bootstrap-slider.min.js"></script>
+    <script src="./static/js/external/jquery-2.2.4.min.js"></script>
```
Those are old versions of the libraries. Also, why are 4 lines added instead of just 2?

rht commented:
I do think the refactor should happen first. It is simpler to pass in urlpath as a Python argument than Bash env.

Corvince commented:
If this is still actively followed please re-open the PR at https://github.com/projectmesa/mesa-viz-tornado
But note that we are moving towards a new pure python frontend
